### PR TITLE
Enable universal checkpointing

### PR DESCRIPTION
**Pull request by tjruwase**
Enable universal checkpointing for GPT implementation. To our knowledge, this is the second model that integrates this feature after the [BLOOM ](https://github.com/bigscience-workshop/bigscience/blob/master/train/tr11-176B-ml/README.md#checkpoint-reshaping)model. Tested with ZeRO stage 1 and fp16. 
Companion DS PR: https://github.com/microsoft/DeepSpeed/pull/4516

- [x] Code cleanup
- [x] Tensorboard screenshot of successful 3D reshaping. 
- [x] bf16 tests
- ~~[ ] ZeRO stage 2 tests~~: Defer Z2 support to separate PR.

### CODE REVIEW

**Comment by tjruwase** [comment]
@stas00, FYI

File: megatron/checkpointing.py [hunk 1]
```diff
@@ -736,3 +754,56 @@ def load_biencoder_checkpoint(model, only_query_model=False,
...
+def _checkpoint_info():
+    args = get_args()
+    tokenizer = get_tokenizer()
+
+    return {
+        "padded_vocab_size": args.padded_vocab_size,
+        "original_vocab_size": tokenizer.vocab_size,
+    }
+
+def _universal_checkpoint_info():
```
Comment (mosheisland) [inline megatron/checkpointing.py:768]: The VOCABULARY_PARAMETER_PATTERNS, PARAMETER_TO_AVERAGE_PATTERNS, PARAMETER_TO_AVERAGE_PATTERNS, PARAMETER_WITH_ROW_PARALLELISM_PATTERNS is model dependent.
Therefore, suggest the following:
- pass model as a parameter to this function (model is derived from MegatronModule)
- MegatronModule will implement "pure-virtual" methods to retrieve required information (e.g. will return a dict with above information)
- Each model (GPT3, LLaMA, Bloom) will have to implement the method(s) to retrieve their specific information

File: megatron/checkpointing.py [hunk 2]
```diff
@@ -736,3 +754,56 @@ def load_biencoder_checkpoint(model, only_query_model=False,
...
+    ]
+
+    # Replicated (shared) parameters on the pipeline dimension 
+    info[PIPELINE_REPLICATED_PARAMETER_PATTERNS] = [
+        r"tied_modules.embed.word_embeddings.weight", 
+        r"tied_modules.embed.position_embeddings.weight"
+    ]
+
+    # Parameter slices that should be averaged not concatenated. 
+    info[PARAMETER_TO_AVERAGE_PATTERNS] = [
```
Comment (mosheisland) [inline megatron/checkpointing.py:789]: I think that Averaging may hide a real issue.
The parameters in this list are expected to be identical across all TP workers.
Therefore, I would rename it to TP_REPLICATED_PATTERNS

File: megatron/model/language_model.py [hunk 3]
```diff
@@ -360,9 +360,13 @@ def forward(self, inputs, **kwargs):
 
     @property
     def word_embeddings_weight(self):
-        """Easy accessory for the DeepSpeed pipeline engine to tie embeddings across stages."""
+        """Easy accessory for the DeepSpeed pipeline engine to tie word embeddings across stages."""
         return self.word_embeddings.weight
 
+    @property
+    def position_embeddings_weight(self):
```
Comment (mosheisland) [inline megatron/model/language_model.py:367]: Bloom, LLaMA do not have position embeddings.

File: megatron/checkpointing.py (hunk 1 above)
Comment (tjruwase) [reply megatron/checkpointing.py:768]: Fair point. Since megatron code supports multiple models, it makes sense to push this logic to the model implementations.

File: megatron/checkpointing.py (hunk 2 above)
Comment (tjruwase) [reply megatron/checkpointing.py:789]: Agreed.

**Comment by tjruwase** [comment]
With 3D parallel training of a toy model with fp16 and zero stage 1, we observe that PP and DP reshape work perfectly, while TP reshaping seems broken. Fixing TP reshaping is future work. 

NLAYERS=4, SEQLEN=512, HIDDEN=512, GLOBAL BATCH SIZE =16, MICRO BATCH SIZE = 4
1. Train with TP=PP=DP=2 for 200 iterations and save checkpoint at iteration 100.
2. Convert iteration 100 checkpoint to universal format. 
3. Resume training at iteration 100 universal checkpoint into the following 3D shapes:
     |  PP  |  TP  | DP  | Note | Loss@Iter101 | 
    | ---   | ---  | ---   |  ---  |  ---  |  
    |   1   |   1   |   1   | Shrink all |   7.964   | 
    |   2   |   2   |   1   | Shrink DP | 7.548 | 
    |   2   |   1   |   2   | Shrink PP | 7.55 |   
    |   1   |   2   |   2   | Shrink TP | 7.962 | 
    |   2   |   2   |   4   | Expand DP | 7.55 |  
    |   2   |   4   |   2   | Expand PP | 7.551 | 
    |   4   |   2   |   2   | Expand TP | 15.31 | 
    |   2   |   1   |   1   | Shrink PP & DP | 7.547 |  
    |   2   |   2   |   2   | No change |  7.547 | 

Below, we share a tensorboard screenshot to compare loss at iteration 101. The original training run is labelled with suffix **toy_save**. We observe that, except for TP reshapes, all the reshapes are identical to original run loss of **7.547**. 
  
image

File: examples_deepspeed/universal_checkpointing/run_fp16.sh
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/run_fp16.sh:1]: Can we add a small README (or even just comments at the top of these scripts) that indicate how these are supposed to be used for testing universal checkpointing?

File: megatron/training.py [hunk 4]
```diff
@@ -615,6 +632,8 @@ def setup_model_and_optimizer(model_provider_func,
     else:
         model[0].global_steps = student_global_steps
 
+    # quit()
+
```
Comment (mrwyattii) [inline megatron/training.py:636]:
```suggestion
```

File: megatron/training.py [hunk 5]
```diff
@@ -641,6 +660,9 @@ def train_step(forward_step_func, data_iterator,
     args = get_args()
     timers = get_timers()
 
+    # from .utils import dump_weights, dump_position_embed_weights
+    # dump_position_embed_weights("before", args.iteration, model)
+
```
Comment (mrwyattii) [inline megatron/training.py:665]:
```suggestion
```

File: megatron/utils.py [hunk 6]
```diff
@@ -286,3 +286,100 @@ def checkpoint_throughput_calculator(model, latency_second):
...
+    #     if mpu.is_pipeline_last_stage()):
+    #         x = optimizer.fp32_groups_flat_partition[1]
+    #         fh.write(f"fp32={x[-402432:]}\n")
+
+    # import os
+    # import socket
+    # hostname = socket.gethostname()
+    # pid = os.getpid()
+    # global_rank = torch.distributed.get_rank()
+    #fn = f"debug-{iteration}-pp{pp_rank}-tp{tp_rank}-dp{dp_rank}-global{global_rank}-{preamble}-{pid}.txt"
```
Comment (mrwyattii) [inline megatron/utils.py:385]: Do we have reason to keep this commented code around? If so, let's add a comment describing why it may be useful in the future.

**Review by mrwyattii — APPROVED** [review]

**Comment by tjruwase** [comment]
Corresponding bf16 results to the above. This used bf16_optimizer (equivalent to zero stage 1) and fp32 gradient accumulation. 

image
